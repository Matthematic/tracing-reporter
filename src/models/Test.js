
class Test {
    constructor({id, name, link, issues, shortLink, type}) {
        this.id = id;
        this.name = name;
        this.link = link;
        this.issues = issues;
        this.shortLink = shortLink;
        this.type = type;
    }

    print() {
        if (this.id || this.name || this.type) {
            console.log(`${this.id} - ${this.name} ${this.type} ${this.issues.toString()}`);
        }
    }
}

export default Test;