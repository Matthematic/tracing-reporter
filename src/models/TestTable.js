import { CONSTANTS } from './Config';

/**
 * Represents a single table of tests mapped to a requirement ID
 * 
 * e.g. "1000000": [
    {
      "name": "AA",
      "link": "../demos/Blackbox/testBlackbox.js#L3",
      "issues": [
        "TRACE-1001",
        "TRACE-1002"
      ],
      "shortLink": "testBlackbox.js",
      "type": "Blackbox"
    },
    ...
  ]
 */
class Table {
    constructor(id) {
      this.id = id;
      this.tests = [];
      this.count = this.tests.length;
    }

    // intended for testing
    clone(table) {
      this.id = table.id;
      this.tests = JSON.parse(JSON.stringify(table.tests));
      return this;
    }

    getTests() {
      return this.tests;
    }

    setId(x) {
      this.id = x;
      return this;
    }

    add(tests) {
      if (Array.isArray(tests)) {
        this.tests = this.tests.concat(tests)
      }
      else {
        this.tests.push(tests);
      }
      this.count = this.tests.length;
      return this;
    }

    sort(key = 'id', direction = CONSTANTS.ASCENDING) {
      this.tests.sort((a, b) => (b.name === a.name) ? 0 : (b.name < a.name) ? 1 : -1) // sort by name first
      this.tests.sort((a, b) => {
          let x = a[key];
          let y = b[key];

          if (Array.isArray(a[key])) { // issues are an array so we need to stringify
            x = a[key].toString();
            y = b[key].toString();
          }

          if (x === y) {
            return 0;
          }

          const compare = direction === CONSTANTS.DESCENDING ? (y > x) : (y < x);
          return compare ? 1 : -1
        });
      return this;
    }

    count() {
      return this.tests.length;
    }

    print() {
      this.tests.forEach(t => t.print());
    }

}

export default Table;