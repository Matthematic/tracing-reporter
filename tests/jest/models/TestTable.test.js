import TestTable from "../../../src/models/TestTable";
import Test from '../../../src/models/Test';

const unitTest = new Test({ id: '1', name: 'Unit Test', link: '"../testUnit.js#L3",', issues: ['TEST-123'], shortLink: 'testUnit.js' , type: 'Unit' })
const grayboxTest = new Test({ id: '2', name: 'Graybox Test', link: '"../testGraybox.js#L3",', issues: ['TEST-223'], shortLink: 'testGraybox.js' , type: 'Graybox' })
const blackboxTest = new Test({ id: '3', name: 'Blackbox Test', link: '"../testBlackbox.js#L3",', issues: ['TEST-323'], shortLink: 'testBlackbox.js' , type: 'Blackbox' })

describe("getTests", () =>  {
  it('returns tests', () => {
    const table = new TestTable().add(unitTest);
    expect(table.getTests()).toEqual([unitTest])
  })
});

describe("id", () =>  {
  it('sets the ID', () => {
    const table = new TestTable().setId('1234567');
    expect(table.id).toEqual('1234567')
  });
});

describe("add", () =>  {
  it('adds single test', () => {
    const table = new TestTable().add(unitTest);
    expect(table.tests).toEqual([unitTest])
  })

  it('adds array of tests', () => {
    const table = new TestTable().add([unitTest, grayboxTest, blackboxTest]);
    expect(table.tests).toEqual([unitTest, grayboxTest, blackboxTest])
  })
});

describe("sort", () =>  {
  ['issues', 'link', 'name', 'type'].forEach(key => {
    it(`sorts by ${key}`, () => {
      const table = new TestTable().add([unitTest, grayboxTest, blackboxTest]);
      expect(table.sort(key).tests).toMatchSnapshot();
    })
  })
});

describe("count", () =>  {
  it('returns test count', () => {
    const table = new TestTable().add([unitTest, grayboxTest, blackboxTest]);
    expect(table.count()).toEqual(3);
  })
});