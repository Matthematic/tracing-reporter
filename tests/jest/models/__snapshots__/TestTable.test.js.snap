// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sort sorts by issues 1`] = `
Array [
  Test {
    "id": "1",
    "issues": Array [
      "TEST-123",
    ],
    "link": "\\"../testUnit.js#L3\\",",
    "name": "Unit Test",
    "shortLink": "testUnit.js",
    "type": "Unit",
  },
  Test {
    "id": "2",
    "issues": Array [
      "TEST-223",
    ],
    "link": "\\"../testGraybox.js#L3\\",",
    "name": "Graybox Test",
    "shortLink": "testGraybox.js",
    "type": "Graybox",
  },
  Test {
    "id": "3",
    "issues": Array [
      "TEST-323",
    ],
    "link": "\\"../testBlackbox.js#L3\\",",
    "name": "Blackbox Test",
    "shortLink": "testBlackbox.js",
    "type": "Blackbox",
  },
]
`;

exports[`sort sorts by link 1`] = `
Array [
  Test {
    "id": "3",
    "issues": Array [
      "TEST-323",
    ],
    "link": "\\"../testBlackbox.js#L3\\",",
    "name": "Blackbox Test",
    "shortLink": "testBlackbox.js",
    "type": "Blackbox",
  },
  Test {
    "id": "2",
    "issues": Array [
      "TEST-223",
    ],
    "link": "\\"../testGraybox.js#L3\\",",
    "name": "Graybox Test",
    "shortLink": "testGraybox.js",
    "type": "Graybox",
  },
  Test {
    "id": "1",
    "issues": Array [
      "TEST-123",
    ],
    "link": "\\"../testUnit.js#L3\\",",
    "name": "Unit Test",
    "shortLink": "testUnit.js",
    "type": "Unit",
  },
]
`;

exports[`sort sorts by name 1`] = `
Array [
  Test {
    "id": "3",
    "issues": Array [
      "TEST-323",
    ],
    "link": "\\"../testBlackbox.js#L3\\",",
    "name": "Blackbox Test",
    "shortLink": "testBlackbox.js",
    "type": "Blackbox",
  },
  Test {
    "id": "2",
    "issues": Array [
      "TEST-223",
    ],
    "link": "\\"../testGraybox.js#L3\\",",
    "name": "Graybox Test",
    "shortLink": "testGraybox.js",
    "type": "Graybox",
  },
  Test {
    "id": "1",
    "issues": Array [
      "TEST-123",
    ],
    "link": "\\"../testUnit.js#L3\\",",
    "name": "Unit Test",
    "shortLink": "testUnit.js",
    "type": "Unit",
  },
]
`;

exports[`sort sorts by type 1`] = `
Array [
  Test {
    "id": "3",
    "issues": Array [
      "TEST-323",
    ],
    "link": "\\"../testBlackbox.js#L3\\",",
    "name": "Blackbox Test",
    "shortLink": "testBlackbox.js",
    "type": "Blackbox",
  },
  Test {
    "id": "2",
    "issues": Array [
      "TEST-223",
    ],
    "link": "\\"../testGraybox.js#L3\\",",
    "name": "Graybox Test",
    "shortLink": "testGraybox.js",
    "type": "Graybox",
  },
  Test {
    "id": "1",
    "issues": Array [
      "TEST-123",
    ],
    "link": "\\"../testUnit.js#L3\\",",
    "name": "Unit Test",
    "shortLink": "testUnit.js",
    "type": "Unit",
  },
]
`;
